### DISCUSSION TITLE
Netherlands national checksum validation

### DISCUSSION THREAD

**Discussion by PabloAJomer** [mdomke/schwifty#205] [General]
Hi, I'm curious where the original source for the national checksum validation for the Netherlands is? 
I'm having trouble finding reliable sources but we have an IBAN that we have tried paying money to and it has worked, that is not valid according to Schwifty. Also I find some sources saying that there are two validation schemes for NL BBAN.

I don't find this source below that creditable but it seems to indicate that there are two validation algorithms of which one is the one implemented by the Schwifty.

https://www.credit-card.be/BankAccount/ValidationRules.htm#NL_Validation

Kindly advice.

**Comment by PabloAJomer** [comment]
I have some further proof. There used to be a so called "elfproef" rule which states to use the alg that is used. This may still be a valid check for some accounts but for newer account this rule seems to not be enforced.

Trying this IBAN: NL08BNPA8026000002 (which we have managed to pay to) we see that the below validator fails but it also states that it's normal to have the elfproof fail. So I suppose it becomes a bit sematic wthere it should be used or not? I haven't figured out if there is a way of knowing if it should be validated or not.

https://www.ibancalculator.com/iban_validieren.html

**Reply by Natim** [reply]
My version of schwifty doesn't raise for it:

```
>>> schwifty.IBAN("NL08BNPA8026000002").validate()
True
```

**Comment by Natim** [comment]
You are right, it's interesting

![image](https://github.com/mdomke/schwifty/assets/229453/e0976369-d037-454d-a265-67e2512039c1)

**Comment by Natim** [comment]
@sh4dowb  since you implemented the validation rules for Netherlands, would you have more context about that?

**Reply by sh4dowb** [reply]
I think there's still a checksum for new accounts, because before implementing it, I had created a new NL IBAN, replaced digits in BBAN (and fixed IBAN checksum) but that generated IBAN failed on some financial websites.
I found the checksum algorithm from a few sources, and all IBANs generated using that algorithm has passed

I'm not sure what is it called, but it appears it's not same with elfproef,
I don't have the original source where I found it from, but just did some searches, maybe worth taking a look at Oracle's docs here:
https://docs.oracle.com/cd/E18727_01/doc.121/e13483/T359831T498954.htm

![image](https://github.com/mdomke/schwifty/assets/12783710/1bfe3e00-9432-48ee-bb25-e956b69cbd71)

I guess the credit-card[.]be page is wrong somehow

btw, the provided BNP IBAN above passes validation on schwifty for me, also just got some real IBANs to double check and they pass as well

**Reply by PabloAJomer** [reply]
I got the following email from their official paymentassociarion (info@betaalvereniging.nl).

```
For starters, we can formally confirm this: you should not verify the BBAN within a Dutch IBAN with the 'elfproef'.
As of 2020 the BBAN within a Dutch IBAN can be any string of 10 digits, with the exception of BBANs that are reserved for G-accounts (BBANs starting with the digits 099).
See: https://www.betaalvereniging.nl/betaalproducten-en-diensten/iban/ibannext/
Kind regards,
Dutch Payments Association

www.paymentsassociation.nl
```

**Reply by PabloAJomer** [reply]
I think it looks a lot like the elfproef algorithm in the code. elfproef basically means 11 test and as we can see in the code we multiply the 10 account digits with the vector and then compare to mod 11.

**Reply by PabloAJomer** [reply]
Also the fact that the bank issued us this iban where we could even make payments speaks in favour of something being wrong since it isn't accepted by the library.

**Reply by PabloAJomer** [reply]
It passes the IBAN validation but not when enabling BBAN validation just to be clear.
```
IBAN("NL08BNPA8026000002", validate_bban=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pablojomer/.pyenv/versions/3.11.5/envs/service-sepa-camt-handler/lib/python3.11/site-packages/schwifty/iban.py", line 77, in __init__
    self.validate(validate_bban)
  File "/Users/pablojomer/.pyenv/versions/3.11.5/envs/service-sepa-camt-handler/lib/python3.11/site-packages/schwifty/iban.py", line 180, in validate
    self.bban.validate_national_checksum()
  File "/Users/pablojomer/.pyenv/versions/3.11.5/envs/service-sepa-camt-handler/lib/python3.11/site-packages/schwifty/bban.py", line 145, in validate_national_checksum
    raise exceptions.InvalidBBANChecksum("Invalid national checksum")
schwifty.exceptions.InvalidBBANChecksum: Invalid national checksum
```

**Reply by mdomke** [reply]
@PabloAJomer 🤔 After reading https://www.betaalvereniging.nl/betaalproducten-en-diensten/iban/ibannext/ it seems to me that we should stop enforcing the checksum validation for Dutch IBANs, even though the `validate_bban` parameter has been set to `True`.

**Reply by PabloAJomer** [reply]
@mdomke Yes, I agree with that assessment. I don't know if there are cases where the "elfprooef" could still be useful but the administrative authority says that it should not be checked any more.
